content/ch-quantum-hardware/Jaynes-Cummings-model.ipynb

```diff
@@ -0,0 +1,764 @@
+   "outputs": [],
+   "source": [
+    "from scipy.optimize import curve_fit\n",
+    "from scipy.signal import savgol_filter\n",
+    "\n",
+    "# samples need to be multiples of 16 to accommodate the hardware limitations\n",
+    "def get_closest_multiple_of_16(num):\n",
+    "    return int(num + 8 ) - (int(num + 8 ) % 16)\n",
+    "\n",
+    "# process the reflective measurement results\n",
+    "# in a reflective measurement the data is encoded in the phase of the output signal\n",
+    "def process_reflective_measurement(freqs,values):\n",
+    "    phase_grad = np.gradient(savgol_filter(np.unwrap(np.angle(values)),3,2),freqs)\n",
+    "    return (phase_grad-min(phase_grad))/(max(phase_grad)-min(phase_grad)) - 1\n",
+    "\n",
+    "# lorentzian function\n",
+    "def lorentzian(f, f0, k, a, offs):\n",
+    "    return -a*k/(2*np.pi)/((k/2)**2+(f-f0)**2)+offs\n",
+    "\n",
+    "# fit a lorentizan function\n",
+    "def fit_lorentzian(freqs,values):\n",
+    "    p0=[freqs[np.argmin(values)],(freqs[-1]-freqs[0])/2,min(values),0]\n",
+    "    bounds=([freqs[0],0,-np.inf,-np.inf],[freqs[-1],freqs[-1]-freqs[0],np.inf,np.inf])\n",
+    "    popt,pcov=curve_fit(lorentzian, freqs, values, p0=p0, bounds=bounds)\n",
+    "    return popt,pcov\n",
+    "\n",
+    "# exponential function\n",
+    "def exponential(t,tau,a,offset):\n",
+    "    return a*np.exp(-t/tau)+offset\n",
+    "\n",
+    "# fit an exponential function\n",
+    "def fit_exponential(ts,values):\n",
+    "    p0=[np.average(ts),1,0]\n",
+    "    return curve_fit(exponential, ts, values,p0=p0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1. Measuring $\\kappa$"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Photons decay out of imperfect electromagnetic cavities. The decay rate $\\kappa$ for the resonator cavity can be measured by calculating the linewidth of the resonance peak in a resonator spectroscopy scan. Larger values of $\\kappa$ imply that the resonator cavity is more lossy. The resonator loss can be quantified using the quality factor $Q=\\omega_r/\\kappa$; higher $Q$ indicates a lower rate of energy loss from the cavity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: DeprecationWarning: the pulse_lib module is deprecated, pulse_lib is renamed to library\n",
+      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
+     ]
+    }
+   ],
+   "source": [
+    "from qiskit import pulse            # This is where we access all of our Pulse features!\n",
+    "from qiskit.pulse import Play, Acquire\n",
+    "from qiskit.pulse import pulse_lib\n",
+    "import numpy as np\n",
+    "\n",
+    "dt=backend_config.dt   # AWG resolution\n",
```

Comment:
What does AWG mean? I don't think it's explained in this chapter

Reply:
Arbitrary waveform generator

Reply:
If it's not mentioned in the chapter, I think it should be included

Reply:
That's a rabbit hole that I don't theink we should climb down in this chapter ... I replace the term AWG with Hardware